riscv32imac no-std support for wasmtime

```
error: Wasmtime is being compiled for an architecture that it does not support. If this architecture is one you would like to see supported you may file an issue on Wasmtime's issue tracker: https://github.com/byt
ecodealliance/wasmtime/issues/new
  --> C:\Users\theo\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasmtime-21.0.1\src\runtime\vm\arch\mod.rs:24:9
   |
24 | /         compile_error!(
25 | |             "Wasmtime is being compiled for an architecture \
26 | |              that it does not support. If this architecture is \
27 | |              one you would like to see supported you may file an \
28 | |              issue on Wasmtime's issue tracker: \
29 | |              https://github.com/bytecodealliance/wasmtime/issues/new\
30 | |         ");
   | |__________^

error[E0432]: unresolved import `core::sync::atomic::AtomicU64`
   --> C:\Users\theo\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasmtime-21.0.1\src\runtime\store.rs:104:5
    |
104 | use core::sync::atomic::AtomicU64;
    |     ^^^^^^^^^^^^^^^^^^^^---------
    |     |                   |
    |     |                   help: a similar name exists in the module: `AtomicU32`
    |     no `AtomicU64` in `sync::atomic`

error[E0432]: unresolved import `core::sync::atomic::AtomicU64`
 --> C:\Users\theo\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasmtime-21.0.1\src\runtime\store\data.rs:8:26
  |
8 | use core::sync::atomic::{AtomicU64, Ordering::Relaxed};
  |                          ^^^^^^^^^
  |                          |
  |                          no `AtomicU64` in `sync::atomic`
  |                          help: a similar name exists in the module: `AtomicU32`
  |
  = help: consider importing this unresolved item through its public re-export instead:
          crate::store::AtomicU64

error[E0432]: unresolved import `core::sync::atomic::AtomicU64`
  --> C:\Users\theo\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasmtime-21.0.1\src\runtime\vm.rs:10:26
   |
10 | use core::sync::atomic::{AtomicU64, AtomicUsize, Ordering};
   |                          ^^^^^^^^^

...

error[E0425]: cannot find value `NEXT_OLDER_FP_FROM_FP_OFFSET` in module `arch`
   --> C:\Users\theo\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasmtime-21.0.1\src\runtime\vm\traphandlers\backtrace.rs:248:63
    |
248 |             let next_older_fp = *(fp as *mut usize).add(arch::NEXT_OLDER_FP_FROM_FP_OFFSET);
    |                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in `arch`

error[E0425]: cannot find function `reached_entry_sp` in module `arch`
   --> C:\Users\theo\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasmtime-21.0.1\src\runtime\vm\traphandlers\backtrace.rs:249:22
    |
249 |             if arch::reached_entry_sp(next_older_fp, trampoline_sp) {
    |                      ^^^^^^^^^^^^^^^^ not found in `arch`

error[E0277]: the trait bound `u128: From<&v128::V128>` is not satisfied
   --> C:\Users\theo\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasmtime-21.0.1\src\runtime\externals\global.rs:190:63
    |
190 |                 Val::V128(i) => *definition.as_u128_mut() = i.into(),
    |                                                               ^^^^ the trait `From<&v128::V128>` is not implemented for `u128`, which is required by `&v128::V128: Into<_>`
    |
    = help: the following other types implement trait `From<T>`:
              <u128 as From<Ipv6Addr>>
              <u128 as From<bool>>
              <u128 as From<char>>
              <u128 as From<core::ascii::Char>>
              <u128 as From<u16>>
              <u128 as From<u32>>
              <u128 as From<u64>>
              <u128 as From<u8>>
              <u128 as From<v128::V128>>
    = note: required for `&v128::V128` to implement `Into<u128>`
...
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -570,6 +570,10 @@ jobs:
           apt_packages: gcc-powerpc64le-linux-gnu
           env:
             CARGO_TARGET_POWERPC64LE_UNKNOWN_LINUX_GNU_LINKER: powerpc64le-linux-gnu-gcc
+        # A no_std target without 64-bit atomics
+        - target: riscv32imac-unknown-none-elf
+          os: ubuntu-latest
+          test: cargo check -p wasmtime --no-default-features --features runtime,gc,component-model
     env: ${{ matrix.env || fromJSON('{}') }}
     steps:
     - uses: actions/checkout@v4

--- a/crates/environ/src/builtin.rs
+++ b/crates/environ/src/builtin.rs
@@ -40,6 +40,7 @@ macro_rules! foreach_builtin_function {
             // Invoked when fuel has run out while executing a function.
             out_of_gas(vmctx: vmctx) -> bool;
             // Invoked when we reach a new epoch.
+            #[cfg(target_has_atomic = "64")]
             new_epoch(vmctx: vmctx) -> u64;
             // Invoked before malloc returns.
             #[cfg(feature = "wmemcheck")]

--- a/crates/wasmtime/src/engine.rs
+++ b/crates/wasmtime/src/engine.rs
@@ -8,6 +8,7 @@ use crate::runtime::vm::GcRuntime;
 use crate::sync::OnceLock;
 use crate::Config;
 use alloc::sync::Arc;
+#[cfg(target_has_atomic = "64")]
 use core::sync::atomic::{AtomicU64, Ordering};
 #[cfg(any(feature = "cranelift", feature = "winch"))]
 use object::write::{Object, StandardSegment};
@@ -61,7 +62,7 @@ struct EngineInner {
     profiler: Box<dyn crate::profiling_agent::ProfilingAgent>,
     #[cfg(feature = "runtime")]
     signatures: TypeRegistry,
-    #[cfg(feature = "runtime")]
+    #[cfg(all(feature = "runtime", target_has_atomic = "64"))]
     epoch: AtomicU64,
 
     /// One-time check of whether the compiler's settings, if present, are
@@ -130,7 +131,7 @@ impl Engine {
                 profiler: config.build_profiler()?,
                 #[cfg(feature = "runtime")]
                 signatures: TypeRegistry::new(),
-                #[cfg(feature = "runtime")]
+                #[cfg(all(feature = "runtime", target_has_atomic = "64"))]
                 epoch: AtomicU64::new(0),
                 #[cfg(any(feature = "cranelift", feature = "winch"))]
                 compatible_with_native_host: OnceLock::new(),
@@ -317,6 +318,9 @@ impl Engine {
         }
         if !cfg!(has_virtual_memory) && self.tunables().memory_init_cow {
             return Err("virtual memory disabled at compile time -- cannot enable CoW".into());
+        }
+        if !cfg!(target_has_atomic = "64") && self.tunables().epoch_interruption {
+            return Err("epochs currently require 64-bit atomics".into());
         }
         Ok(())
     }
@@ -692,10 +696,12 @@ impl Engine {
         self.config().custom_code_memory.as_ref()
     }
 
+    #[cfg(target_has_atomic = "64")]
     pub(crate) fn epoch_counter(&self) -> &AtomicU64 {
         &self.inner.epoch
     }
 
+    #[cfg(target_has_atomic = "64")]
     pub(crate) fn current_epoch(&self) -> u64 {
         self.epoch_counter().load(Ordering::Relaxed)
     }
@@ -725,6 +731,7 @@ impl Engine {
     /// This method is signal-safe: it does not make any syscalls, and
     /// performs only an atomic increment to the epoch value in
     /// memory.
+    #[cfg(target_has_atomic = "64")]
     pub fn increment_epoch(&self) {
         self.inner.epoch.fetch_add(1, Ordering::Relaxed);
     }

--- a/crates/wasmtime/src/runtime/component/resources.rs
+++ b/crates/wasmtime/src/runtime/component/resources.rs
@@ -11,7 +11,7 @@ use core::fmt;
 use core::marker;
 use core::mem::MaybeUninit;
 use core::ptr::NonNull;
-use core::sync::atomic::{AtomicU64, Ordering::Relaxed};
+use core::sync::atomic::{AtomicU32, Ordering::Relaxed};
 use wasmtime_environ::component::{
     CanonicalAbiInfo, ComponentTypes, DefinedResourceIndex, InterfaceType, ResourceIndex,
     TypeResourceTableIndex,
@@ -248,12 +248,18 @@ pub struct Resource<T> {
 ///   borrow tracking associated with it. The low 32-bits of the value are
 ///   the table index and the upper 32-bits are the generation.
 ///
-/// Note that this is an `AtomicU64` but it's not intended to actually be
-/// used in conjunction with threads as generally a `Store<T>` lives on one
-/// thread at a time. The `AtomicU64` here is used to ensure that this type
-/// is `Send + Sync` when captured as a reference to make async programming
-/// more ergonomic.
-struct AtomicResourceState(AtomicU64);
+/// Note that this is two `AtomicU32` fields but it's not intended to actually
+/// be used in conjunction with threads as generally a `Store<T>` lives on one
+/// thread at a time. The pair of `AtomicU32` here is used to ensure that this
+/// type is `Send + Sync` when captured as a reference to make async
+/// programming more ergonomic.
+///
+/// Also note that two `AtomicU32` here are used instead of `AtomicU64` to be
+/// more portable to platforms without 64-bit atomics.
+struct AtomicResourceState {
+    index: AtomicU32,
+    generation: AtomicU32,
+}
 
 #[derive(Debug, PartialEq, Eq, Copy, Clone)]
 enum ResourceState {
@@ -264,39 +270,52 @@ enum ResourceState {
 }
 
 impl AtomicResourceState {
-    const BORROW: Self = Self(AtomicU64::new(ResourceState::BORROW));
-    const NOT_IN_TABLE: Self = Self(AtomicU64::new(ResourceState::NOT_IN_TABLE));
+    const BORROW: Self = AtomicResourceState::new(ResourceState::Borrow);
+    const NOT_IN_TABLE: Self = AtomicResourceState::new(ResourceState::NotInTable);
+
+    const fn new(state: ResourceState) -> AtomicResourceState {
+        let (index, generation) = state.encode();
+        Self {
+            index: AtomicU32::new(index),
+            generation: AtomicU32::new(generation),
+        }
+    }
 
     fn get(&self) -> ResourceState {
-        ResourceState::decode(self.0.load(Relaxed))
+        ResourceState::decode(self.index.load(Relaxed), self.generation.load(Relaxed))
     }
 
     fn swap(&self, state: ResourceState) -> ResourceState {
-        ResourceState::decode(self.0.swap(state.encode(), Relaxed))
+        let (index, generation) = state.encode();
+        let index_prev = self.index.load(Relaxed);
+        self.index.store(index, Relaxed);
+        let generation_prev = self.generation.load(Relaxed);
+        self.generation.store(generation, Relaxed);
+        ResourceState::decode(index_prev, generation_prev)
     }
 }
 
 impl ResourceState {
     // See comments on `state` above for info about these values.
-    const BORROW: u64 = u64::MAX;
-    const NOT_IN_TABLE: u64 = u64::MAX - 1;
-    const TAKEN: u64 = u64::MAX - 2;
+    const BORROW: u32 = u32::MAX;
+    const NOT_IN_TABLE: u32 = u32::MAX - 1;
+    const TAKEN: u32 = u32::MAX - 2;
 
-    fn decode(bits: u64) -> ResourceState {
-        match bits {
+    fn decode(idx: u32, generation: u32) -> ResourceState {
+        match generation {
             Self::BORROW => Self::Borrow,
             Self::NOT_IN_TABLE => Self::NotInTable,
             Self::TAKEN => Self::Taken,
-            other => Self::Index(HostResourceIndex(other)),
+            _ => Self::Index(HostResourceIndex::new(idx, generation)),
         }
     }
 
-    fn encode(&self) -> u64 {
+    const fn encode(&self) -> (u32, u32) {
         match self {
-            Self::Borrow => Self::BORROW,
-            Self::NotInTable => Self::NOT_IN_TABLE,
-            Self::Taken => Self::TAKEN,
-            Self::Index(index) => index.0,
+            Self::Borrow => (0, Self::BORROW),
+            Self::NotInTable => (0, Self::NOT_IN_TABLE),
+            Self::Taken => (0, Self::TAKEN),
+            Self::Index(index) => (index.index(), index.generation()),
         }
     }
 }
@@ -359,12 +378,12 @@ impl HostResourceIndex {
         HostResourceIndex(u64::from(idx) | (u64::from(generation) << 32))
     }
 
-    fn index(&self) -> u32 {
-        u32::try_from(self.0 & 0xffffffff).unwrap()
+    const fn index(&self) -> u32 {
+        (self.0 & 0xffffffff) as u32
     }
 
-    fn generation(&self) -> u32 {
-        u32::try_from(self.0 >> 32).unwrap()
+    const fn generation(&self) -> u32 {
+        (self.0 >> 32) as u32
     }
 }
 

--- a/crates/wasmtime/src/runtime/store.rs
+++ b/crates/wasmtime/src/runtime/store.rs
@@ -222,6 +222,7 @@ pub struct StoreInner<T> {
 
     limiter: Option<ResourceLimiterInner<T>>,
     call_hook: Option<CallHookInner<T>>,
+    #[cfg(target_has_atomic = "64")]
     epoch_deadline_behavior:
         Option<Box<dyn FnMut(StoreContextMut<T>) -> Result<UpdateDeadline> + Send + Sync>>,
     // for comments about `ManuallyDrop`, see `Store::into_data`
@@ -614,6 +615,7 @@ impl<T> Store<T> {
             },
             limiter: None,
             call_hook: None,
+            #[cfg(target_has_atomic = "64")]
             epoch_deadline_behavior: None,
             data: ManuallyDrop::new(data),
         });
@@ -995,6 +997,7 @@ impl<T> Store<T> {
     /// See documentation on
     /// [`Config::epoch_interruption()`](crate::Config::epoch_interruption)
     /// for an introduction to epoch-based interruption.
+    #[cfg(target_has_atomic = "64")]
     pub fn set_epoch_deadline(&mut self, ticks_beyond_current: u64) {
         self.inner.set_epoch_deadline(ticks_beyond_current);
     }
@@ -1025,6 +1028,7 @@ impl<T> Store<T> {
     /// See documentation on
     /// [`Config::epoch_interruption()`](crate::Config::epoch_interruption)
     /// for an introduction to epoch-based interruption.
+    #[cfg(target_has_atomic = "64")]
     pub fn epoch_deadline_trap(&mut self) {
         self.inner.epoch_deadline_trap();
     }
@@ -1056,6 +1060,7 @@ impl<T> Store<T> {
     /// See documentation on
     /// [`Config::epoch_interruption()`](crate::Config::epoch_interruption)
     /// for an introduction to epoch-based interruption.
+    #[cfg(target_has_atomic = "64")]
     pub fn epoch_deadline_callback(
         &mut self,
         callback: impl FnMut(StoreContextMut<T>) -> Result<UpdateDeadline> + Send + Sync + 'static,
@@ -1086,7 +1091,7 @@ impl<T> Store<T> {
     /// See documentation on
     /// [`Config::epoch_interruption()`](crate::Config::epoch_interruption)
     /// for an introduction to epoch-based interruption.
-    #[cfg(feature = "async")]
+    #[cfg(all(feature = "async", target_has_atomic = "64"))]
     pub fn epoch_deadline_async_yield_and_update(&mut self, delta: u64) {
         self.inner.epoch_deadline_async_yield_and_update(delta);
     }
@@ -1184,6 +1189,7 @@ impl<'a, T> StoreContextMut<'a, T> {
     /// Sets the epoch deadline to a certain number of ticks in the future.
     ///
     /// For more information see [`Store::set_epoch_deadline`].
+    #[cfg(target_has_atomic = "64")]
     pub fn set_epoch_deadline(&mut self, ticks_beyond_current: u64) {
         self.0.set_epoch_deadline(ticks_beyond_current);
     }
@@ -1191,6 +1197,7 @@ impl<'a, T> StoreContextMut<'a, T> {
     /// Configures epoch-deadline expiration to trap.
     ///
     /// For more information see [`Store::epoch_deadline_trap`].
+    #[cfg(target_has_atomic = "64")]
     pub fn epoch_deadline_trap(&mut self) {
         self.0.epoch_deadline_trap();
     }
@@ -1200,7 +1207,7 @@ impl<'a, T> StoreContextMut<'a, T> {
     ///
     /// For more information see
     /// [`Store::epoch_deadline_async_yield_and_update`].
-    #[cfg(feature = "async")]
+    #[cfg(all(feature = "async", target_has_atomic = "64"))]
     pub fn epoch_deadline_async_yield_and_update(&mut self, delta: u64) {
         self.0.epoch_deadline_async_yield_and_update(delta);
     }
@@ -1920,7 +1927,7 @@ impl StoreOpaque {
     ///
     /// This only works on async futures and stores, and assumes that we're
     /// executing on a fiber. This will yield execution back to the caller once.
-    #[cfg(feature = "async")]
+    #[cfg(all(feature = "async", target_has_atomic = "64"))]
     fn async_yield_impl(&mut self) -> Result<()> {
         use crate::runtime::vm::Yield;
 
@@ -2688,6 +2695,7 @@ unsafe impl<T> crate::runtime::vm::VMStore for StoreInner<T> {
         Ok(())
     }
 
+    #[cfg(target_has_atomic = "64")]
     fn new_epoch(&mut self) -> Result<u64, anyhow::Error> {
         // Temporarily take the configured behavior to avoid mutably borrowing
         // multiple times.
@@ -2769,6 +2777,7 @@ unsafe impl<T> crate::runtime::vm::VMStore for StoreInner<T> {
 }
 
 impl<T> StoreInner<T> {
+    #[cfg(target_has_atomic = "64")]
     pub(crate) fn set_epoch_deadline(&mut self, delta: u64) {
         // Set a new deadline based on the "epoch deadline delta".
         //
@@ -2783,10 +2792,12 @@ impl<T> StoreInner<T> {
         *epoch_deadline = self.engine().current_epoch() + delta;
     }
 
+    #[cfg(target_has_atomic = "64")]
     fn epoch_deadline_trap(&mut self) {
         self.epoch_deadline_behavior = None;
     }
 
+    #[cfg(target_has_atomic = "64")]
     fn epoch_deadline_callback(
         &mut self,
         callback: Box<dyn FnMut(StoreContextMut<T>) -> Result<UpdateDeadline> + Send + Sync>,

--- a/crates/wasmtime/src/runtime/store/data.rs
+++ b/crates/wasmtime/src/runtime/store/data.rs
@@ -5,7 +5,6 @@ use core::fmt;
 use core::marker;
 use core::num::NonZeroU64;
 use core::ops::{Index, IndexMut};
-use core::sync::atomic::{AtomicU64, Ordering::Relaxed};
 
 // This is defined here, in a private submodule, so we can explicitly reexport
 // it only as `pub(crate)`. This avoids a ton of
@@ -210,24 +209,42 @@ impl StoreId {
     /// Allocates a new unique identifier for a store that has never before been
     /// used in this process.
     pub fn allocate() -> StoreId {
-        static NEXT_ID: AtomicU64 = AtomicU64::new(0);
+        use core::sync::atomic::Ordering::Relaxed;
 
-        // Only allow 2^63 stores at which point we start panicking to prevent
-        // overflow.
+        // When 64-bit atomics are allowed then allow 2^63 stores at which point
+        // we start panicking to prevent overflow.
         //
         // If a store is created once per microsecond then this will last the
         // current process for 584,540 years before overflowing.
-        //
-        // Also note the usage of `Relaxed` ordering here which should be ok
+        #[cfg(target_has_atomic = "64")]
+        use core::sync::atomic::AtomicU64 as AtomicId;
+        #[cfg(target_has_atomic = "64")]
+        const OVERFLOW_THRESHOLD: u64 = 1 << 63;
+
+        // When 64-bit atomics are not allowed then for now use 32-bit atomics
+        // instead. That means that if a store is created once per microsecond
+        // then the process will last a little over an hour. That's not a lot
+        // but it's expected that these platforms probably aren't churning out
+        // stores. If such a situation happens the process will at least be torn
+        // down instead of possibly silently corrupting things. For now this is
+        // at least the least invasive way to add support for platforms without
+        // 64-bit atomics at this time.
+        #[cfg(not(target_has_atomic = "64"))]
+        use core::sync::atomic::AtomicU32 as AtomicId;
+        #[cfg(not(target_has_atomic = "64"))]
+        const OVERFLOW_THRESHOLD: u32 = u32::MAX - 100_000;
+
+        // Note the usage of `Relaxed` ordering here which should be ok
         // since we're only looking for atomicity on this counter and this
         // otherwise isn't used to synchronize memory stored anywhere else.
+        static NEXT_ID: AtomicId = AtomicId::new(0);
         let id = NEXT_ID.fetch_add(1, Relaxed);
-        if id & (1 << 63) != 0 {
-            NEXT_ID.store(1 << 63, Relaxed);
+        if id > OVERFLOW_THRESHOLD {
+            NEXT_ID.store(OVERFLOW_THRESHOLD, Relaxed);
             panic!("store id allocator overflow");
         }
 
-        StoreId(NonZeroU64::new(id + 1).unwrap())
+        StoreId(NonZeroU64::new(u64::from(id) + 1).unwrap())
     }
 
     #[inline]

--- a/crates/wasmtime/src/runtime/store/data.rs
+++ b/crates/wasmtime/src/runtime/store/data.rs
@@ -209,40 +209,45 @@ impl StoreId {
     /// Allocates a new unique identifier for a store that has never before been
     /// used in this process.
     pub fn allocate() -> StoreId {
-        use core::sync::atomic::Ordering::Relaxed;
-
         // When 64-bit atomics are allowed then allow 2^63 stores at which point
         // we start panicking to prevent overflow.
         //
         // If a store is created once per microsecond then this will last the
         // current process for 584,540 years before overflowing.
-        #[cfg(target_has_atomic = "64")]
-        use core::sync::atomic::AtomicU64 as AtomicId;
-        #[cfg(target_has_atomic = "64")]
         const OVERFLOW_THRESHOLD: u64 = 1 << 63;
 
-        // When 64-bit atomics are not allowed then for now use 32-bit atomics
-        // instead. That means that if a store is created once per microsecond
-        // then the process will last a little over an hour. That's not a lot
-        // but it's expected that these platforms probably aren't churning out
-        // stores. If such a situation happens the process will at least be torn
-        // down instead of possibly silently corrupting things. For now this is
-        // at least the least invasive way to add support for platforms without
-        // 64-bit atomics at this time.
-        #[cfg(not(target_has_atomic = "64"))]
-        use core::sync::atomic::AtomicU32 as AtomicId;
-        #[cfg(not(target_has_atomic = "64"))]
-        const OVERFLOW_THRESHOLD: u32 = u32::MAX - 100_000;
+        #[cfg(target_has_atomic = "64")]
+        let id = {
+            use core::sync::atomic::{AtomicU64, Ordering::Relaxed};
 
-        // Note the usage of `Relaxed` ordering here which should be ok
-        // since we're only looking for atomicity on this counter and this
-        // otherwise isn't used to synchronize memory stored anywhere else.
-        static NEXT_ID: AtomicId = AtomicId::new(0);
-        let id = NEXT_ID.fetch_add(1, Relaxed);
-        if id > OVERFLOW_THRESHOLD {
-            NEXT_ID.store(OVERFLOW_THRESHOLD, Relaxed);
-            panic!("store id allocator overflow");
-        }
+            // Note the usage of `Relaxed` ordering here which should be ok
+            // since we're only looking for atomicity on this counter and this
+            // otherwise isn't used to synchronize memory stored anywhere else.
+            static NEXT_ID: AtomicId = AtomicId::new(0);
+            let id = NEXT_ID.fetch_add(1, Relaxed);
+            if id > OVERFLOW_THRESHOLD {
+                NEXT_ID.store(OVERFLOW_THRESHOLD, Relaxed);
+                panic!("store id allocator overflow");
+            }
+        };
+
+        // When 64-bit atomics are not allowed use a `RwLock<u64>`. This is
+        // already used elsewhere in Wasmtime and currently has the
+        // implementation of panic-on-contention, but it's at least no worse
+        // than what wasmtime had before and is at least correct and UB-free.
+        #[cfg(not(target_has_atomic = "64"))]
+        let id = {
+            use crate::sync::RwLock;
+            static NEXT_ID: RwLock<u64> = RwLock::new(0);
+
+            let mut lock = NEXT_ID.write();
+            if *lock > OVERFLOW_THRESHOLD {
+                panic!("store id allocator overflow");
+            }
+            let ret = *lock;
+            *lock += 1;
+            ret
+        };
 
         StoreId(NonZeroU64::new(u64::from(id) + 1).unwrap())
     }

--- a/crates/wasmtime/src/runtime/store/data.rs
+++ b/crates/wasmtime/src/runtime/store/data.rs
@@ -223,12 +223,13 @@ impl StoreId {
             // Note the usage of `Relaxed` ordering here which should be ok
             // since we're only looking for atomicity on this counter and this
             // otherwise isn't used to synchronize memory stored anywhere else.
-            static NEXT_ID: AtomicId = AtomicId::new(0);
+            static NEXT_ID: AtomicU64 = AtomicU64::new(0);
             let id = NEXT_ID.fetch_add(1, Relaxed);
             if id > OVERFLOW_THRESHOLD {
                 NEXT_ID.store(OVERFLOW_THRESHOLD, Relaxed);
                 panic!("store id allocator overflow");
             }
+            id
         };
 
         // When 64-bit atomics are not allowed use a `RwLock<u64>`. This is
@@ -249,7 +250,7 @@ impl StoreId {
             ret
         };
 
-        StoreId(NonZeroU64::new(u64::from(id) + 1).unwrap())
+        StoreId(NonZeroU64::new(id + 1).unwrap())
     }
 
     #[inline]

--- a/crates/wasmtime/src/runtime/vm.rs
+++ b/crates/wasmtime/src/runtime/vm.rs
@@ -178,6 +178,7 @@ pub unsafe trait VMStore {
     /// Callback invoked whenever an instance observes a new epoch
     /// number. Cannot fail; cooperative epoch-based yielding is
     /// completely semantically transparent. Returns the new deadline.
+    #[cfg(target_has_atomic = "64")]
     fn new_epoch(&mut self) -> Result<u64, Error>;
 
     /// Callback invoked whenever an instance needs to trigger a GC.

--- a/crates/wasmtime/src/runtime/vm/instance.rs
+++ b/crates/wasmtime/src/runtime/vm/instance.rs
@@ -22,6 +22,7 @@ use core::alloc::Layout;
 use core::any::Any;
 use core::ops::Range;
 use core::ptr::NonNull;
+#[cfg(target_has_atomic = "64")]
 use core::sync::atomic::AtomicU64;
 use core::{mem, ptr};
 use sptr::Strict;
@@ -571,6 +572,7 @@ impl Instance {
     }
 
     /// Return a pointer to the global epoch counter used by this instance.
+    #[cfg(target_has_atomic = "64")]
     pub fn epoch_ptr(&mut self) -> NonNull<Option<VmPtr<AtomicU64>>> {
         unsafe { self.vmctx_plus_offset_mut(self.offsets().ptr.vmctx_epoch_ptr()) }
     }
@@ -596,11 +598,13 @@ impl Instance {
             let store = store.as_mut();
             self.runtime_limits()
                 .write(Some(store.vmruntime_limits().into()));
+            #[cfg(target_has_atomic = "64")]
             self.epoch_ptr()
                 .write(Some(NonNull::from(store.engine().epoch_counter()).into()));
             self.set_gc_heap(store.gc_store_mut().ok());
         } else {
             self.runtime_limits().write(None);
+            #[cfg(target_has_atomic = "64")]
             self.epoch_ptr().write(None);
             self.set_gc_heap(None);
         }

--- a/crates/wasmtime/src/runtime/vm/libcalls.rs
+++ b/crates/wasmtime/src/runtime/vm/libcalls.rs
@@ -1074,6 +1074,7 @@ fn out_of_gas(store: &mut dyn VMStore, _instance: &mut Instance) -> Result<()> {
 }
 
 // Hook for when an instance observes that the epoch has changed.
+#[cfg(target_has_atomic = "64")]
 fn new_epoch(store: &mut dyn VMStore, _instance: &mut Instance) -> Result<NextEpoch> {
     store.new_epoch().map(NextEpoch)
 }

--- a/crates/wasmtime/src/runtime/vm/provenance.rs
+++ b/crates/wasmtime/src/runtime/vm/provenance.rs
@@ -46,7 +46,7 @@ use core::fmt;
 use core::marker;
 use core::num::NonZeroUsize;
 use core::ptr::NonNull;
-use core::sync::atomic::{AtomicU64, AtomicUsize};
+use core::sync::atomic::AtomicUsize;
 use wasmtime_environ::VMSharedTypeIndex;
 
 /// A pointer that is used by compiled code, or in other words is accessed
@@ -213,7 +213,8 @@ unsafe impl VmSafe for i64 {}
 unsafe impl VmSafe for i128 {}
 unsafe impl VmSafe for isize {}
 unsafe impl VmSafe for AtomicUsize {}
-unsafe impl VmSafe for AtomicU64 {}
+#[cfg(target_has_atomic = "64")]
+unsafe impl VmSafe for core::sync::atomic::AtomicU64 {}
 
 // This is a small `u32` wrapper defined in `wasmtime-environ`, so impl the
 // vm-safe-ness here.

--- a/crates/wasmtime/src/runtime/vm/sys/custom/capi.rs
+++ b/crates/wasmtime/src/runtime/vm/sys/custom/capi.rs
@@ -116,6 +116,7 @@ unsafe extern "C" {
     ///
     /// Returns false if `wasmtime_longjmp` was used to return to this function.
     /// Returns true if `wasmtime_longjmp` was not called and `callback` returned.
+    #[cfg(has_host_compiler_backend)]
     pub fn wasmtime_setjmp(
         jmp_buf: *mut *const u8,
         callback: extern "C" fn(*mut u8, *mut u8) -> bool,
@@ -133,6 +134,7 @@ unsafe extern "C" {
     ///
     /// This function may be invoked from the `wasmtime_trap_handler_t`
     /// configured by `wasmtime_init_traps`.
+    #[cfg(has_host_compiler_backend)]
     pub fn wasmtime_longjmp(jmp_buf: *const u8) -> !;
 
     /// Initializes trap-handling logic for this platform.

--- a/crates/wasmtime/src/runtime/vm/sys/custom/mod.rs
+++ b/crates/wasmtime/src/runtime/vm/sys/custom/mod.rs
@@ -17,6 +17,7 @@ pub mod capi;
 #[cfg(has_virtual_memory)]
 pub mod mmap;
 pub mod traphandlers;
+#[cfg(has_host_compiler_backend)]
 pub mod unwind;
 #[cfg(has_virtual_memory)]
 pub mod vm;

--- a/crates/wasmtime/src/runtime/vm/sys/custom/traphandlers.rs
+++ b/crates/wasmtime/src/runtime/vm/sys/custom/traphandlers.rs
@@ -1,13 +1,14 @@
 use crate::prelude::*;
-use crate::runtime::vm::VMContext;
-use core::mem;
-use core::ptr::NonNull;
+#[cfg(has_host_compiler_backend)]
+use crate::{mem, ptr::NonNull, runtime::vm::VMContext};
 
+#[cfg(has_host_compiler_backend)]
 pub use crate::runtime::vm::sys::capi::{self, wasmtime_longjmp};
 
 #[allow(missing_docs)]
 pub type SignalHandler = Box<dyn Fn() + Send + Sync>;
 
+#[cfg(has_host_compiler_backend)]
 pub unsafe fn wasmtime_setjmp(
     jmp_buf: *mut *const u8,
     callback: extern "C" fn(*mut u8, NonNull<VMContext>) -> bool,

--- a/crates/wasmtime/src/runtime/vm/sys/custom/traphandlers.rs
+++ b/crates/wasmtime/src/runtime/vm/sys/custom/traphandlers.rs
@@ -1,6 +1,8 @@
 use crate::prelude::*;
 #[cfg(has_host_compiler_backend)]
-use crate::{mem, ptr::NonNull, runtime::vm::VMContext};
+use crate::runtime::vm::VMContext;
+#[cfg(has_host_compiler_backend)]
+use core::{mem, ptr::NonNull};
 
 #[cfg(has_host_compiler_backend)]
 pub use crate::runtime::vm::sys::capi::{self, wasmtime_longjmp};

--- a/docs/stability-tiers.md
+++ b/docs/stability-tiers.md
@@ -94,12 +94,15 @@ For explanations of what each tier means see below.
 | Target               | `armv7-unknown-linux-gnueabihf`   | full-time maintainer |
 | Target               | `i686-pc-windows-msvc`            | CI testing, full-time maintainer |
 | Target               | `i686-unknown-linux-gnu`          | full-time maintainer |
+| Target               | `powerpc64le-unknown-linux-gnu`   | CI testing, full-time maintainer |
+| Target               | `riscv32imac-unknown-none-elf`[^5]| CI testing, full-time maintainer |
 | Target               | `riscv64gc-unknown-linux-gnu`     | full-time maintainer        |
 | Target               | `wasm32-wasip1` [^3]              | Supported but not tested    |
 | Target               | `x86_64-linux-android`            | CI testing, full-time maintainer |
 | Target               | `x86_64-unknown-freebsd`          | CI testing, full-time maintainer |
 | Target               | `x86_64-unknown-illumos`          | CI testing, full-time maintainer |
 | Target               | `x86_64-unknown-linux-musl` [^4]  | CI testing, full-time maintainer |
+| Target               | `x86_64-unknown-none` [^5]        | CI testing, full-time maintainer |
 | Compiler Backend     | Winch on x86\_64                  | WebAssembly proposals (`simd`, `relaxed-simd`, `tail-call`, `reference-types`, `threads`)     |
 | Compiler Backend     | Winch on aarch64                  | Complete implementation     |
 | Execution Backend    | Pulley                            | fuzzing                     |
@@ -138,6 +141,12 @@ Wasmtime-compiled-to-wasm can itself compile wasm but cannot execute wasm.
 linked, meaning that they are not suitable for "run on any linux distribution"
 style use cases. Wasmtime does not have static binary artifacts at this time and
 that will require building from source.
+
+[^5]: Rust targets that are `#![no_std]` don't support the entire feature set of
+Wasmtime. For example the `threads` Cargo feature requires the standard library.
+For more information see the [`no_std` documentation][nostd].
+
+[nostd]: ./stability-platform-support.md
 
 #### Unsupported features and platforms
 

